File: 2020/day01/day01.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type pair struct {
	a, b int
}

type triplet struct {
	a, b, c int
}

func pairWithSum(numbers []int, sum int) (pair, error) {
	others := make(map[int]bool)
	for i := 1; i < len(numbers); i++ {
		others[numbers[i]] = true
	}
	for i := 0; i < len(numbers)-1; i++ {
		if others[sum-numbers[i]] {
			return pair{numbers[i], sum - numbers[i]}, nil
		}
	}
	return pair{0, 0}, errors.New("Pair not found")
}

func tripletWithSum(numbers []int, sum int) (triplet, error) {
	sort.Ints(numbers)
	for i := 0; i < len(numbers)-2; i++ {
		for j, k := i+1, len(numbers)-1; j < k; {
			total := numbers[i] + numbers[j] + numbers[k]
			// fmt.Printf("Trying %d + %d + %d (= %d)\n", numbers[i], numbers[j], numbers[k], total)
			if total == sum {
				return triplet{numbers[i], numbers[j], numbers[k]}, nil
			} else if total < sum {
				j++
			} else {
				k--
			}
		}
	}
	return triplet{0, 0, 0}, errors.New("Triplet not found")
}

func loadNumbers(fn string) ([]int, error) {
	file, err := os.Open(fn)
	defer file.Close()
	if err != nil {
		return nil, err
	}

	var numbers = make([]int, 200)
	scanner := bufio.NewScanner(file)
	var index = 0
	for scanner.Scan() {
		number, err := strconv.Atoi(scanner.Text())
		if err != nil {
			return nil, err
		}
		if err := scanner.Err(); err != nil {
			return nil, err
		}

		numbers[index] = number
		index++
	}

	return numbers, nil
}

func main() {
	numbers, err := loadNumbers("day01_input.txt")
	if err != nil {
		log.Fatal(err)
	}

	pair, err := pairWithSum(numbers, 2020)
	if err == nil {
		fmt.Printf("%d\n", pair.a*pair.b)
	} else {
		fmt.Println(err)
	}

	triplet, err := tripletWithSum(numbers, 2020)
	if err == nil {
		fmt.Printf("%d\n", triplet.a*triplet.b*triplet.c)
	} else {
		fmt.Println(err)
	}
}

File: 2020/day01/day01.ts
const fs = require('fs');

type Pair = [number, number];
type Triplet = [number, number, number];

const pairWithSum = (numbers: number[], sum: number): Pair => {
  let others: Set<number> = new Set();
  for (let i = 1; i < numbers.length; i++) {
    others.add(numbers[i]);
  }
  for (let i = 0; i < numbers.length - 1;i++) {
    if (others.has(sum - numbers[i])) {
      return [numbers[i], sum - numbers[i]];
    }
  }
  throw new Error(`No pair with sum ${sum} found`);
}

const tripletWithSum = (numbers: number[], sum: number): Triplet => {
  numbers.sort((a, b) => a - b);
  for (let i = 0; i < numbers.length - 2; i++) {
    let j = i + 1;
    let k = numbers.length - 1;
    while (j < k) {
      const total = numbers[i] + numbers[j] + numbers[k];
      if (total === sum) {
        return [numbers[i], numbers[j], numbers[k]];
      } else if (total < sum) {
        j++;
      } else {
        k--;
      }
    }
  }
  throw new Error(`No triplet with sum ${sum} found`);
}

const loadNumbers = (): number[] =>
  fs.readFileSync('./2020/day01/day01_input.txt', 'utf8').split("\n").map(l => parseInt(l));

const part1Result = (): number => {
  const pair = pairWithSum(loadNumbers(), 2020);
  return pair[0] * pair[1];

}

const part2Result = (): number => {
  const triplet = tripletWithSum(loadNumbers(), 2020);
  return triplet[0] * triplet[1] * triplet[2];
}

export { part1Result, part2Result }
